File: dotnet/Allmark/Render/ConsoleInsertionRenderer.cs
namespace Allmark.Render;

using Allmark.Types;

public static class ConsoleInsertionRenderer
{
    public static Renderer Create(string style, string reset)
    {
        return new Renderer
        {
            Name = "insertion",
            Render = (node, state, first, last, decode) => Render(node, state, style, reset),
        };
    }

    public static Renderer Create()
    {
        return new Renderer
        {
            Name = "insertion",
            Render = (node, state, first, last, decode) => Render(node, state, RenderToConsole.AnsiGreen, RenderToConsole.AnsiReset),
        };
    }

    public static void Render(MarkdownNode node, RendererState state, string style, string reset)
    {
        state.Output.Append($"{style}++");
        RenderChildren.Execute(node, state);
        state.Output.Append($"++{reset}");
    }
}

File: zig/src/render-console/consoleInsertionRenderer.zig
const std = @import("std");

const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const ConsoleRendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const renderChildrenConsole = @import("console.zig").renderChildrenConsole;

pub const consoleInsertionRenderer = Renderer{
    .name = "insertion",
    .render = render,
};

pub fn render(node: *const MarkdownNode, state: *ConsoleRendererState, first: ?bool, last: ?bool, decode: ?bool) void {
    _ = first;
    _ = last;
    _ = decode;

    state.output.appendSlice(state.allocator, "++") catch unreachable;
    renderChildrenConsole(node, state, true) catch unreachable;
    state.output.appendSlice(state.allocator, "++") catch unreachable;
}
